prometheus_juju_exporter/exporter.py

```diff
@@ -16,17 +16,48 @@ def __init__(self, debug=False):
         self.logger = get_logger(debug=debug)
         self.logger.info("Parsed config: %s", self.config.config_dir())
         self._registry = CollectorRegistry()
-        self.collector = CollectorDaemon(self._registry, debug=debug)
+        self.metrics = {}
+        self.collector = CollectorDaemon(debug=debug)
         self.logger.debug("Exporter initialized")
 
+    def _create_metrics_dict(self, gauge_name, gauge_desc, labels=[]):
+        """Create a dict of gauge instances."""
+        if gauge_name not in self.metrics:
+            self.logger.debug("Creating Gauge %s", gauge_name)
+            self.metrics[gauge_name] = Gauge(
+                gauge_name, gauge_desc, labelnames=labels, registry=self._registry
+            )
+
+    def update_registry(self, data):
+        for gauge_name in data.keys():
+            self._create_metrics_dict(
+                gauge_name=gauge_name,
+                gauge_desc=data[gauge_name]["gauge_desc"],
+                labels=data[gauge_name]["labels"],
+            )
+            for labels in data[gauge_name]["labelvalues_update"]:
+                value = labels.pop("value")
+                self.logger.debug(
+                    "Updating Gauge %s, %s: %s", gauge_name, labels, value
+                )
+                self.metrics[gauge_name].labels(**labels).set(value)
+
+            for labels in data[gauge_name]["labelvalues_remove"]:
+                labels.pop()
+                self.logger.debug(
+                    "Deleting labelvalues %s from %s...", labels, gauge_name
+                )
+                self.metrics[gauge_name].remove(*labels)
+
     async def trigger(self, **kwargs):
         """Configure prometheus_client gauges from generated stats."""
         run_collector = True
 
         while run_collector:
```

Comment:
I had problems testing code with this implementation. Maybe we can have a discussion about this tomorrow. Nevertheless, since this is a future improvement idea, this could be a part of a different PR.

Reply:
I opened an issue #8 so we can keep track of this idea.